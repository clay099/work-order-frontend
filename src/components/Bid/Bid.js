import React from "react";
import useFields from "../../hooks/useFields";
import { TextField, Button, TableCell } from "@material-ui/core";
import { useDispatch, useSelector, shallowEqual } from "react-redux";
import { Send } from "@material-ui/icons/";
import { submitNewBidToAPI, updateBidWithAPI } from "../../actions/bids";
import useToggle from "../../hooks/useToggle";

const Bid = ({ projectId }) => {
	const { token, bid } = useSelector((st) => {
		let bid;
		// to avoid typeError checks if the projectId has a bid attached
		if (!st.bids[projectId]) {
			// returns undefined if no bid for projectId
			bid = undefined;
		} else {
			// returns value if project does have a bid.
			// note the user might not have placed a bid yet and result can still be undefined
			bid = st.bids[projectId][st.login.id];
		}

		return {
			token: st.login.token,
			bid,
		};
	}, shallowEqual);

	const { toggle: disabled, setTrue, setFalse } = useToggle(Boolean(bid));

	const INITIALSTATE = { bid: bid || "" };
	const { formData, handleChange } = useFields(INITIALSTATE);
	const dispatch = useDispatch();

	const handleSubmit = async (e) => {
		e.preventDefault();
		// Add submit logic
		if (!bid) {
			await dispatch(
				submitNewBidToAPI({ token, projectId, bid: +formData.bid })
			);
		} else {
			await dispatch(
				updateBidWithAPI({ token, projectId, bid: +formData.bid })
			);
		}
		setTrue();
	};

	return (
		<>
			<TableCell size="small">
				<TextField
					required
					autoFocus
					id="bid"
					label="Bid"
					type="number"
					name="bid"
					onChange={handleChange}
					value={formData.bid}
					size="small"
					disabled={disabled}
				></TextField>
			</TableCell>
			<TableCell>
				{!disabled ? (
					<Button
						type="submit"
						size="small"
						variant="contained"
						color="primary"
						endIcon={<Send />}
						onClick={handleSubmit}
					>
						{bid ? "Update " : "Place "}
						Bid
					</Button>
				) : (
					<Button
						type="submit"
						size="small"
						variant="outlined"
						color="primary"
						endIcon={<Send />}
						onClick={setFalse}
					>
						Edit Bid
					</Button>
				)}
			</TableCell>
		</>
	);
};

export default Bid;
