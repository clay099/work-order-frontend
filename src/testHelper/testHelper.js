// testHelper.js
import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore } from "redux";
import { Provider } from "react-redux";
// Import your own reducer
import reducer from "../reducers/rootReducer";

function render(
	ui,
	{
		initialState,
		store = createStore(reducer, initialState),
		...renderOptions
	} = {}
) {
	function Wrapper({ children }) {
		return <Provider store={store}>{children}</Provider>;
	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const tokenState = { token: "token" };
const userTypeUserState = { userType: "user" };
const userTypeTradesmenState = { userType: "user" };
const idState = { id: 1234 };
const userDetailsState = { userDetails: "userDetails" };
const handleSubmitState = { handleSubmit: jest.fn() };
const handleChangeState = { handleChange: jest.fn() };
const disabledState = { disabled: false };
const disabledEmailState = { disabledEmail: false };
const buttonTextState = { buttonText: "buttonText" };
const formDataState = {
	formData: {
		id: 1,
		firstName: "firstName",
		lastName: "lastName",
		email: "email",
		password: "password",
		newPassword: "newPassword",
		confirmNewPassword: "confirmNewPassword",
	},
};

// re-export everything
export * from "@testing-library/react";
// override render method
export {
	render,
	tokenState,
	userTypeUserState,
	userTypeTradesmenState,
	idState,
	userDetailsState,
	handleSubmitState,
	handleChangeState,
	disabledState,
	disabledEmailState,
	buttonTextState,
	formDataState,
};
