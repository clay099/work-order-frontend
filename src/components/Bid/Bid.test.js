import React from "react";
import Bid from "./Bid";
import { render } from "../../testHelper/testHelper";
console.error = jest.fn();

const params = {
	projectId: 1,
};

const initialState = {
	login: { id: 1 },
	bids: { 1: { 1: "500" } },
};

describe("Bid Component", () => {
	// smoke test
	test("renders without crashing", () => {
		render(<Bid {...params} />, {
			initialState,
		});
	});

	// snapshot test
	test("matches snapshot", () => {
		const { asFragment } = render(<Bid {...params} />, {
			initialState,
		});
		expect(asFragment).toMatchSnapshot();
	});

	test("value to be in the document", () => {
		const { getByText } = render(<Bid {...params} />, {
			initialState,
		});
		const value = getByText("Bid");
		expect(value).toBeInTheDocument();
	});
});
