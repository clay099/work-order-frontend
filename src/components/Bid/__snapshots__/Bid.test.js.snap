// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bid Component matches snapshot 1`] = `[Function]`;
